### PR TITLE
Chain handlers

### PR DESCRIPTION
**Pull request by crackerplace**
@ruflin 

Refactored the handler logic to use handler chaining.I did not move the handlers to separate file assuming that's something being discussed in parallel.

Not sure if I should include this change in #128.

### CODE REVIEW

**Comment by ruflin** [comment]
@crackerplace For the part of moving the handler to an other file: I think we are all on the same page that it should move to it's own file but during doing this, also the tests should be cleaned up which could be a bigger task then expected.

Thanks for the PR. Will have a look tomoroow.

**Comment by crackerplace** [comment]
@ruflin Having the handlers in the same file should also be fine as we don't have many handlers to deal with and can move once we get there so as to avoid more complexity from the points that you have mentioned.

File: beater/server.go [hunk 1]
```diff
@@ -157,29 +159,18 @@ func reportInfo(code int, err error) (int, error) {
...
 }
 
-// checkSecretToken checks the Authorization header. It must be in the form of:
-//
+// isAuthorized checks the Authorization header. It must be in the form of:
 //   Authorization: Bearer <secret-token>
-//
 // Bearer must be part of it.
-func checkSecretToken(req *http.Request, secretToken string) bool {
-	// No token configured
```
Comment (ruflin) [inline beater/server.go]: Can you bring this comment back?

File: beater/server.go [hunk 2]
```diff
@@ -157,29 +159,18 @@ func reportInfo(code int, err error) (int, error) {
...
+func isAuthorized(req *http.Request, secretToken string) bool {
```
Comment (ruflin) [inline beater/server.go]: I like the renaming as it's more generic.

File: beater/server.go [hunk 3]
```diff
@@ -38,12 +38,9 @@ func newMuxer(config Config, report reporter) *http.ServeMux {
...
-		handler := createHandler(p, config, report)
-
+		handler := appHandler(p, config, report)
 		logp.Info("Path %s added to request handler", path)
-
-		mux.HandleFunc(path, handler)
+		mux.Handle(path, authHandler(config.SecretToken, handler))
 	}
 
 	mux.HandleFunc("/healthcheck", func(w http.ResponseWriter, r *http.Request) {
```
Comment (ruflin) [inline beater/server.go]: This change just made me realise that we don't check the token for the `/healthcheck` handler I think?

File: beater/server.go [hunk 4]
```diff
@@ -90,16 +87,23 @@ func stop(server *http.Server, timeout time.Duration) {
...
+			sendStatus(w, r, 401, errInvalidToken)
+			return
+		}
+		h.ServeHTTP(w, r)
+	})
+}
 
+func appHandler(p processor.Processor, config Config, report reporter) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logp.Debug("handler", "Request: URI=%s, method=%s, content-length=%d", r.RequestURI, r.Method, r.ContentLength)
```
Comment (simitt) [inline beater/server.go]: With this change a request is is not logged any more when it is not authorized.

**Comment by roncohen** [comment]
Nice work @crackerplace. How about adding a handler for counting and one for logging?

File: beater/server.go (hunk 3 above)
Comment (crackerplace) [reply beater/server.go]: This looks fine for me.

File: beater/server.go (hunk 4 above)
Comment (crackerplace) [reply beater/server.go]: yes,I was looking for a better way to log.Could not use reportInfo in sendStatus as I wanted to.

**Comment by crackerplace** [comment]
@roncohen Sure.Added a log handler.Logging the metric counter from the same handler as well.Let me know if you think we need a separate counting handler.

**Comment by crackerplace** [comment]
@ruflin Do we need responseValid ?

I somehow think the request rate and error rate serve the purpose.The other point being tracking all the return points in case we want to account for a successful response and increment responseValid at that point might become tricky.

File: beater/server.go (hunk 1 above)
Comment (crackerplace) [reply beater/server.go]: done

**Pull request merged**